### ISSUE TITLE
How to have more than one shared memory region in pta

### ISSUE THREAD

**Issue by divneil** [OP-TEE/optee_os#2403]
Hi,

I was writing some custom code in PTA to have 2 shared memory regions.
The requirement is to send some buffer out to REE which it interprets and sends back a buffer in response to that.

In trying so, I see that the va from the below code always give the same virtual address with different size.
FILE: `core/arch/arm/tee/pta_socket.c`
```
    va = tee_fs_rpc_cache_alloc(params[1].memref.size, &mobj, &cookie);
    if (!va)
        return TEE_ERROR_OUT_OF_MEMORY;
```
Can you please help me understand what am I missing. Thanks.

**Comment by jenswi-linaro** [comment]
`tee_fs_rpc_cache_alloc()` is used to avoid frequent alloc, do_cmd, free sequences. Both alloc and free results in a trip to normal world user space. If you look at the implementation of `tee_fs_rpc_cache_alloc()` you can see that it tries to reuse last allocation if it was large enough. The buffer is only to be used for the current thread.

However, it only deals with one buffer so if it's not good enough to allocate a large buffer which then is split into several buffers then you'd need to use `thread_rpc_alloc_payload()` and `thread_rpc_free_payload()` instead.

**Comment by divneil** [comment]
Thanks for pretty quick reply. :)

Yeah! I had seen the definition of `tee_fs_rpc_cache_alloc()` and I tested with removing the `if()` check for size comparison (just to see the behavior), still va was same.
So, internally, `tee_fs_rpc_cache_alloc()`  calls `thread_rpc_alloc_payload()`. I will try it again to see if I did something wrong last time.

Hmm, I have nothing more to ask right away, so, closing the ticket for now.
